beacon_node/eth2-libp2p/src/rpc/handler.rs

```diff
@@ -397,103 +401,107 @@ where
-                                                RPCError::ErrorResponse(error_code),
-                                            ));
-                                        }
-                                        // send the response
-                                        // if it's a single rpc request or an error, close the stream after
-                                        *substream_state =
-                                            InboundSubstreamState::ResponsePendingSend {
-                                                substream: substream,
-                                                message: response,
-                                                closing: !res_is_multiple | res_is_error, // close if an error or we are not expecting more responses
-                                            };
-                                    }
-                                }
-                            }
-                            InboundSubstreamState::ResponsePendingSend {
-                                substream,
-                                message,
-                                closing,
-                            } if res_is_multiple => {
-                                // the stream is in use, add the request to a pending queue
-                                self.queued_outbound_items
-                                    .entry(rpc_id)
-                                    .or_insert_with(Vec::new)
-                                    .push(response);
-
-                                // return the state
-                                *substream_state = InboundSubstreamState::ResponsePendingSend {
-                                    substream,
-                                    message,
-                                    closing,
-                                };
-                            }
-                            InboundSubstreamState::ResponsePendingFlush { substream, closing }
-                                if res_is_multiple =>
-                            {
-                                // the stream is in use, add the request to a pending queue
-                                self.queued_outbound_items
-                                    .entry(rpc_id)
-                                    .or_insert_with(Vec::new)
-                                    .push(response);
-
-                                // return the state
-                                *substream_state = InboundSubstreamState::ResponsePendingFlush {
-                                    substream,
-                                    closing,
-                                };
-                            }
-                            InboundSubstreamState::Closing(substream) => {
+                let (substream_state, protocol, inbound_id) = match rpc_id {
+                    None => return, // TODO: verify this logic. What happens with a Pong response?
```

Comment:
@AgeManning what do you think about this? (We are in `inject_event` in the rpc handler, receiving a `RPCEvent::Response(rpc_id/ *this is an option*/, response)` This line is a response without an Id, with inbound substreams identified by a RequestId (a number, not an option. Since they are inbound, I asume they must have an Id regardless of type)

Reply:
Unless I'm missing something, As you've done, we should have different types for RequestId's and ResponseIds. A response must match a request, so therefore shouldn't be an option. But we can make requests that we don't care about the specific response. Therefore, I think a RequestId should be an option and a ResponseId must have a number. 

In this case, this None variant shouldn't exist right?

Reply:
Are you suggesting to have rpc responses always with an ID regardless of direction or that here this line should never happen but maintaining the responses with optional ids? 

I'm not sure since (from what I understand) responses are in either direction so we can be sending one or receiving one. So wouldn't a ping that we request have a None Id so that the pong we get doesn't have a request id either?

This is the case of responses inside the inbound handling so that's why I think the return is OK (maybe a log actually)

Reply:
Ah yep, my heads been out of the RPC for a bit. Lets go through the cases:
- RPC Inbound Request: Someone requests something for us: Stream gets tagged with an `InboundRequestId` must be `Some`
- RPC Inbound Response: We respond to the request, should use the InboundRequestId to respond, must be `Some`. 
- RPC Outbound Request: We make a request. Can be `None`. The stream still gets some sort of `stream_id` when stored in a mapping in the handler. 
- RPC Outbound Response: A response was received to one of our requests: It will be associated with the `stream_id` of our request, but at the application layer, we don't care about it. 

The question you have here is about the second one right? We are sending a response to an inbound request. I guess in this case, it should never be none, and we should throw an error. 

I think the confusion here is that I've tried to use two types to represent the four different types of logic. We could make four types, but it's a bit work. We could just make the Id's options and throw errors on the cases that shouldn't happen. Which way would you like to go with this PR?

Reply:
It took me a while to check it but I think this way it's fine. I'm just logging the error in this particular case (the second case indeed). I'll clean up and mark as ready for review :+1: